source/common/router/header_parser.h

```diff
@@ -14,54 +14,52 @@
 namespace Envoy {
 namespace Router {
 
+class HeaderParserBase {
+protected:
+  void addHeaders(Http::HeaderMap& headers, const AccessLog::RequestInfo& request_info) const;
+
+  void
+  setHeadersToAdd(const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers);
+
+  std::vector<std::pair<Http::LowerCaseString, HeaderFormatterPtr>> header_formatters_;
+};
+
 class RequestHeaderParser;
 typedef std::unique_ptr<RequestHeaderParser> RequestHeaderParserPtr;
 
 /**
- * This class holds the parsing logic required during configuration build and
- * also perform evaluation for the variables at runtime.
+ * This class provides request-time generation of upstream request headers. Header configurations
+ * are pre-parsed to select between constant values and values based on the evaluation of
+ * AccessLog::RequestInfo fields.
  */
-class RequestHeaderParser {
+class RequestHeaderParser : public HeaderParserBase {
 public:
-  virtual ~RequestHeaderParser() {}
-
   static RequestHeaderParserPtr
   parse(const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers);
 
   void evaluateRequestHeaders(Http::HeaderMap& headers,
                               const AccessLog::RequestInfo& request_info) const;
-
-private:
-  std::list<std::pair<Http::LowerCaseString, HeaderFormatterPtr>> header_formatters_;
-
-  static HeaderFormatterPtr parseInternal(const std::string& format, const bool append);
 };
 
 class ResponseHeaderParser;
 typedef std::unique_ptr<ResponseHeaderParser> ResponseHeaderParserPtr;
 
 /**
- * This class holds the logic required to apply response headers additions and removals.
+ * This class provides request-time manipulation of response headers. Header configurations are
+ * pre-parsed to select between constant values and values based on the evaluation of
+ * AccessLog::RequestInfo fields.
  */
-class ResponseHeaderParser {
+class ResponseHeaderParser : public HeaderParserBase {
```

Comment:
Same here: I think instead of inheriting from HeaderParserBase, it would work just as well to have a HeaderParserBase as a member variable.